### ISSUE TITLE
Multipart Upload Throughput Approaches Zero For Large Files

### ISSUE THREAD

**Issue by shanebarnes** [minio/minio#7206]
For large file (i.e., > 5GB) multipart uploads, it seems that as the upload progresses, the overall application throughput decays exponentially. This performance degradation is most noticeable for a large number of parts (i.e., > 500 parts).

<img width="914" alt="image" src="https://user-images.githubusercontent.com/7718600/52454082-35015500-2b18-11e9-80d9-78ceedf3d52c.png">

## Expected Behavior

I am expecting that large multipart uploads should reach a steady-state application throughput that is independent of part size but limited by other "natural" causes (e.g., network- or disk-limited).

## Current Behavior

For example, when uploading a 25GiB file with a single client connection to a Minio server over a dedicated network connection with no significant latency (i.e., < 1ms) the average application throughput varies depending on the part size:

1. `5GiB part size x 5 parts` => 600Mbps,
2. `500MiB part size x 50 parts` => 540Mbps,
3. `50MiB part size x 500 parts` => 540Mbps, and
4. `5MiB part size x 5000 parts` => 16Mbps.

The 5000-part upload mentioned above exhibits significantly degraded performance and the running average application throughput diminishes exponentially over time.

## Possible Solution

There appears to be increasing delay as more object part files are created temporarily while a multipart upload is in progress. Increasing the part size helps to reduce the symptoms but it does not eliminate the cause.

## Steps to Reproduce (for bugs)

1. `mkdir -p /data/bucket`
2. `minio server /data`
3. `aws s3 cp 25GB.bin s3://bucket`

Note: Try step 3 for different part sizes with a single connection to the Minio server.

## Context

For large file uploads using multipart object puts, small part sizes are unusable since the time it takes to upload the file is extremely slow.

## Your Environment

* `uname -a`
	* Linux ip-xx-xx-xx-xxx.us-west-2.compute.internal 4.14.88-88.76.amzn2.x86_64 #1 SMP Mon Jan 7 18:43:26 UTC 2019 x86_64 x86_64 x86_64 GNU/Linux

* `curl http://169.254.169.254/latest/meta-data/instance-type`
	* t2.2xlarge

* `curl http://169.254.169.254/latest/meta-data/ami-id`
	* ami-032509850cf9ee54e

* `minio version`
	* Version: 2018-10-25T01:27:03Z
	* Release-Tag: RELEASE.2018-10-25T01-27-03Z
	* Commit-ID: bf66e9a529fed3d6c460bcdbbd0a60bc167866a4

**Comment by nitisht** [comment]
thanks for reporting @shanebarnes we'll take a look

**Comment by harshavardhana** [comment]
// cc @krishnasrinivas

**Comment by krishnasrinivas** [comment]
only way to fix it is to have env flag which switches to pwrite() instead of background append. We need to discuss if it should be done. So users can choose full S3 compatibility Vs performance.

**Comment by kannappanr** [comment]
@shanebarnes You can also start minio in `local erasure coded` mode instead of `fs` mode.

`minio server /data1 /data2 /data3 /data4`

**Comment by shanebarnes** [comment]
@kannappanr I am restricted to a single drive in my use of minio.

**Comment by kannappanr** [comment]
@shanebarnes You can still use a single drive but use four folders on a single drive.

`minio server /disk1/folder1 /disk1/folder2 /disk3/folder3 /disk4/folder4`

**Comment by shanebarnes** [comment]
@kannappanr I do need the object parts re-assembled into a single file once the upload is complete so I think I still need `fs` mode. Also, I tried `local erasure coded` mode with 4 directories and it did help but the decaying behavior was still present.

**Comment by harshavardhana** [comment]
How did try erasure coding , do you have multiple disks locally ?

**Comment by kannappanr** [comment]
@harshavardhana He has only one disk.

**Comment by harshavardhana** [comment]
Then trying erasure coding won't help. You need multiple disks.

The skewed number is expected for FS if you use Multipart. You need to upload a longer stream - we support a single stream upto 5TiB so if you do PUT without multipart upload we will avoid background append, this will avoid double writes.

On an NVMe like disk this is not a problem, but especially if you're using ebs volumes not sure how much performance drop is expected for double writes.

So what you are seeing is expected. Either we append later or append in background performance limitations of Multipart API on FS exists.

In erasure coding we don't do that and parts are kept as is, so there is really no drop in performance. Of course given that you are using multiple disks ..

**Comment by shanebarnes** [comment]
@harshavardhana The multipart upload is more of a necessity when I upload over LFNs. Would you expect the exponentially decaying upload rate with FS multipart to be eliminated if I replaced the background append with [os.WriteAt()/pwrite()](https://golang.org/pkg/os/#File.WriteAt)? I am just trying to figure out if I can overcome the problem if I make the changes to my local copy of the source.

**Comment by harshavardhana** [comment]
Background append cannot do pwrite because Multipart upload allows part in any order and can be varying in size i.e each part has its own size.

Unless we can strictly disallow varying part sizes and also disallow unordered uploads it's possible. But this means an S3 incompatible implementation.. our current approach full fills all the needs under best case scenarios. So if you are uploading 200Gig files then your app should be designed such that it uploads parts in order i.e sequentially one part follows the other starting with first part.  The only reason you would see degradation is when parts come in unordered fashion.

**Comment by shanebarnes** [comment]
The degradation I plotted above is only with a single connection with no latency, so I don't think the parts are arriving out-of-order. I can eliminate the EBS volume and try something else. The EBS volume has a 128MB/sec write limit and I would be happy with an average upload rate of 64MB/sec or less (given the double write) as long as it is more or less constant. But the upload rate decays as the uncommitted part list increases. I have also tested the EBS volume with a 25GB FTP transfer just to ensure that a 128MB/sec write speed is achievable and sustainable.

Thanks very much for taking the time to address my questions.

**Comment by harshavardhana** [comment]
> The degradation I plotted above is only with a single connection with no latency, so I don't think the parts are arriving out-of-order. I can eliminate the EBS volume and try something else. The EBS volume has a 128MB/sec write limit and I would be happy with an average upload rate of 64MB/sec or less (given the double write) as long as it is more or less constant. But the upload rate decays as the uncommitted part list increases. I have also tested the EBS volume with a 25GB FTP transfer just to ensure that a 128MB/sec write speed is achievable and sustainable.

With this demonstration, we need access to the code or tool to replicate it... One potential problem which I didn't state before is that background append doesn't directly take the call for a multipart request. It is further appended after the multipart is successfully saved on the backend. So it means there is basically a double write. Meaning there is a write-in-progress to create a single contiguous file.

**Comment by shanebarnes** [comment]
Here are some more details on my setup. I am currently testing with two EC2 instances in the same data center. I have also reproduced this behavior on my MacBook Pro with a SSD in the past but I am currently running low on available disk space.

__Tools__
* `AWS CLI `: aws-cli/1.15.80 Python/2.7.14 Linux/4.14.88-88.76.amzn2.x86_64 botocore/1.10.79 (pre-installed)
* `Client Machine`: t2.2xlarge + 512GiB gp2 + Amazon Linux 2 in Oregon data center
* `Minio`: 2019-02-06T21:16:36Z | RELEASE.2019-02-06T21-16-36Z | 817269475fa2984754834fab1cdd380d99007dbc
* `Server Machine`: t2.2xlarge + 512GiB gp2 + Amazon Linux 2 in Oregon data center

__AWS CLI Config (~/.aws/config)__
```
[default]
region = us-east-1
s3 =
    multipart_threshold = 5MB
    multipart_chunksize = 5MB
    max_concurrent_requests = 1
    addressing_style = path
```

Note that if I reduce the total number of parts required for the upload by changing the chunk size above to `50MB` or `100MB`, the upload performance is significantly better for a 25GiB file both in terms of speed and stability. 

__Client Commands__

```
# Create large file
$ dd if=/dev/urandom of=25GiB.bin bs=5M count=5000
# Multipart upload using HTTP
$ aws s3 cp 25GiB.bin s3://<bucket> --endpoint-url http://<server ip>
```

__Server Commands__

`minio server /data`

**Comment by harshavardhana** [comment]
AWS CLI chunks 5MiB per part and ends up uploading in parallel, and it is unordered. You can clearly observe this with --debug so no wonder you would see the degradation as you increase the file size. This is what I was indicating previously.

That is why mc splits at 64MiB by default and chooses optimal size automatically based on the file size. 

In a way you are simply hitting the limitation of FS backend due to Multipart expected behavior and our compliance to this behavior.

**Comment by shanebarnes** [comment]
I don't think the AWS CLI is uploading parts in parallel when configured for a request concurrency of 1 (see `max_concurrent_requests = 1` which I posted above), which is how I configured it. I have also increased the AWS CLI part size to 50MB and see similar behavior, although it takes longer to manifest itself. I have also tried this with my own client written using the Go AWS SDK and I have observed the same behavior. In both cases, parts are being uploaded in order across a single HTTP connection and there is a decaying throughput as the uncommitted part list increases.

I can try the same test with `mc` and get back to you.

Multipart uploads with a connection concurrency greater than 1 are not very useful across LFNs if parts have to arrive on the server side in order. I understand from your comments this is a requirement of `fs` mode based on the implementation. I do question if part size can be adequately optimized without considering network latency and connection concurrency. In my experience with both Azure Blob and AWS S3, the disadvantage of using larger part sizes across networks with significant BDPs becomes apparent when a part must be retried or where the part size is significantly larger than the TCP connection's send window.

**Comment by harshavardhana** [comment]
In our experience bigger part size is not a problem for S3 either. Even a retry is not a problem with 64MiB part size with modern networks. I can understand if the part size is like 1G it can become a problem which is not needed.  You can go upto 525MiB per part and upload a 5TiB file.

It is correct there is a balance too, now if the number parts increase which are uploaded quickly i.e the upload speed v/s the write to another contiguous file then we can also see this problem not just parallelly uploaded parts.

When we do our tests we usually use an NVMe, where increase in part size or backlogs do not build up so much even with concurrent uploads. Write speeds are pretty high compared to overall inbound speed (network)

TBH it's a limitation which cannot be avoided for our S3 compliance and FS mode is doing the best it can. 

Unless filesystems provide us a mechanism other than relying on POSIX API limitations or if you are okay having a custom format where parts stay as is and not assembled together - this behavior is here to stay.

**Comment by challarao** [comment]
Is it a regression after September release?

**Comment by harshavardhana** [comment]
No it is not a regression.

**Comment by ghost** [comment]
Yes, I can definitely concurr that there is an issue with minio server.
```
rclone copy -P --s3-chunk-size=5M --s3-upload-concurrency=1 /tmp/samplefile minio:test/
```
Elapsed time:     30m0.4s
while
```
rclone copy -P --s3-chunk-size=15M --s3-upload-concurrency=1 /tmp/samplefile minio:test/
```
Elapsed time:     5m20.8s
We have some difference here, I already brought it up on that community portal there, but on that web-IRC no one really cares.

Some here say that it might be a problem with concurrency
```
rclone copy -P --s3-chunk-size=5M --s3-upload-concurrency=1000 /tmp/samplefile minio:test/
```
Elapsed time:       8m50s
Well, that's not the case. And my tests are clear that it does not matter if the samples come in in the same order or not. It is just that when CPU hits the 100% mark, transfer gets slower and it never recovers, as is nicely shown on the graph up there in the initial issue description.
However, a loose append:
```
cat /dev/urandom | rclone rcat -P --s3-chunk-size=5M --s3-upload-concurrency=1 minio:test/samplefile
```
is pretty steady and does not appear to have this issue, that's why I also tried  –s3-disable-checksum but that did not help either.
```
rclone copy -P --s3-disable-checksum --s3-chunk-size=5M --s3-upload-concurrency=1 /tmp/samplefile minio:test/
```
with "checksum":{"algorithm":"","blocksize":0,"hashes":null}
goes down in bandwidth pretty soon.

**Comment by ghost** [comment]
Now to the interpretation. We can rule out that it has anything to do with the concurrency, instead we can definitely proove that it increases the CPU load the longer it runs.

Proof being:

Hit it with 1000 chunks at once and it will immediately complete (going LINESPEED Gigabit!!!, OK, it will still need a minute to transfer 5 GB but that is looking immediate to me after spending multiple hours on waiting 30 minutes for the transfer to get stuck) with almost no CPU load.
Yes, and then it will take 8 minutes to put the file together, but whatever...

That means that there may be somewhere a while true loop which calculates the MD5 hash, or does something CPU intensive as long as the transfer is ongoing. 

The more chunks come in, the more it has to do and WHEN it reaches a point that calculation of one list (one loop run) takes longer than new ones come in, we have a problem.
More simply said, the new chunks arrive faster than the CPU can complete its list calculation or whatever it does.
Then the CPU is just too slow to catch up and the transfer rate goes to 0.

It also has nothing to do with HDD/SSD/NVMe write speeds as stated above, because
1. we are not assembling (concatenating) anything yet, that happens at the end of the transfer (and yes, that takes time, but that's not the issue here ... I do not care if the "transfer" takes 2 or 5 mins on 100% - not CPU load but completeness as seen by the client, here rclone - meaning, the transfer is complete, but has not finished yet, it is stuck on 100% - instead maybe on 99%, but whatever - because the server has to put files together first - and yes, there is IO load at the end of transfer, but NOT during the transfer)
1. while transfering I am looking at the writes with ZFS zpool iostat -v 1 and I see the average rate of writing to the disc as being approximately the same as the inbound traffic to minio, so there is NO double writing anywhere, except at the end when he has to put the chunks together into the complete file (which is a nothing burger non-issue to me).

**Comment by ghost** [comment]
Well and finally comes some idology.

Some here say that "yes, it is not a problem, as long as you use bigger chunk sizes" and so on.

I do not like this talk at all because it defies the purpose of S3.

The big advantage of S3 is the ability to transfer in chunks.
Alternatively you could take some kind of messaging system, like Kafka, but that would require you to have that big Kafka installation standing somewhere with good bandwidth connectivity, which is not cheap.

Instead you go for S3 and put in 5MB chunks. And that you can do for a long time, normally.
Tried that with digitalocean spaces when they came out and it worked well.

btw. does anyone know what they use as backend? Do they have Ceph or sth else?

By that you can have a microservice running on a small instance, like a digitalocean smallest droplet in order to do even more propaganda, and it runs cheaply. Now imagine you would need to cache 64 MB chunks or sth like that. Either you needed much more RAM or at least a lot of disk writes with that. On a droplet RAM is scarce and with multiple connections à 64 MB you are out of RAM pretty fast, so the only remaining is disk space, which can become a bottleneck on those virtual instances also pretty fast.
S3 has the big advantage to take all this problems/considerations away. Finally if you write on disk, you could then take a blockdevice anyway and re-size it according to needs, instead of using S3 altogether.

**Comment by jonathan-molyneux** [comment]
This problem is also affecting products like cloudberry when doing image based backups.
It uses 1MB blocks (I think) and during the multipart upload (especially if it's resumed) results in excessive cpu usage in minio as the hash requirements increase.
Eventually resulting in a stalemate and uploads cease.

Had anyone found a way to mitigate the issue server side ?
As in this use case it's not possible to accommodate another block size.

**Comment by frittentheke** [comment]
Is this still an issue or already resolved somehow?

**Comment by darkdragon-001** [comment]
@frittentheke Without extensive testing, it still appears to be an issue in my setup... I would really appreciate if someone could look into this.

**Comment by darkdragon-001** [comment]
I think I identified the problem:

`PutObjectPartHandler()` calls `ListObjectParts(.., 0, 1, ...)` to get meta data (`li.UserDefined`) from the first part (start at `0`, max `1` part):
https://github.com/minio/minio/blob/7dbfea1353fc292d20cd399305edf00e4b91f105/cmd/object-handlers.go#L2080-L2085

Even though parameters `partNumberMarker=0` and `maxParts=1`, `ListObjectParts()` processes the whole directory:

https://github.com/minio/minio/blob/7dbfea1353fc292d20cd399305edf00e4b91f105/cmd/fs-v1-multipart.go#L383-L387

and only removes unneeded results in the end:

https://github.com/minio/minio/blob/7dbfea1353fc292d20cd399305edf00e4b91f105/cmd/fs-v1-multipart.go#L436-L444

Instead, it would be enough to read the meta file directly:

https://github.com/minio/minio/blob/7dbfea1353fc292d20cd399305edf00e4b91f105/cmd/fs-v1-multipart.go#L469-L481

@harshavardhana

**Comment by harshavardhana** [comment]
> I think I identified the problem:
> 
> `PutObjectPartHandler()` calls `ListObjectParts(.., 0, 1, ...)` to get meta data (`li.UserDefined`) from the first part (start at `0`, max `1` part):
> https://github.com/minio/minio/blob/7dbfea1353fc292d20cd399305edf00e4b91f105/cmd/object-handlers.go#L2080-L2085
> 
> Even though parameters `partNumberMarker=0` and `maxParts=1`, `ListObjectParts()` processes the whole directory:
> 
> https://github.com/minio/minio/blob/7dbfea1353fc292d20cd399305edf00e4b91f105/cmd/fs-v1-multipart.go#L383-L387
> 
> and only removes unneeded results in the end:
> 
> https://github.com/minio/minio/blob/7dbfea1353fc292d20cd399305edf00e4b91f105/cmd/fs-v1-multipart.go#L436-L444
> 
> Instead, it would be enough to read the meta file directly:
> 
> https://github.com/minio/minio/blob/7dbfea1353fc292d20cd399305edf00e4b91f105/cmd/fs-v1-multipart.go#L469-L481
> 
> @harshavardhana

That is not the problem @darkdragon-001 the problem is the coalesce that needs to be done at the end of multipart on FS mode.

Although ListObjectParts can be a bit optimized is not where the issue is.

**Comment by darkdragon-001** [comment]
@harshavardhana Please read again through the thread (especially https://github.com/minio/minio/issues/7206#issuecomment-505835867). I think agreement was reached that the coalesce at the end (tracked in #9657) is **not** the problem we are looking at here!

When you have a chunk size of 1 MB, a file of 100 GB will have 100 000 chunks. The 1000th chunk (1% process) already has to list a folder with 1000 files. Each consecutive chunk has to list a folder with even more chunks. In total you have to do 1+2+...+100 000 = 5 000 050 000 file lookups. When you can do 1000 file lookups per millisecond (NVMe/PCIe 1 million IOPs, SATA SSD 100 000, SATA HDD <80), this is still 5000 seconds or 1.4 hours (10 000h on SATA HDD). This is just for listing the files, no transfer, checksum or anything else.

**Comment by harshavardhana** [comment]
> When you have a chunk size of 1 MB, a file of 100 GB will have 100 000 chunks. The 1000th chunk (1% process) already has to list a folder with 1000 files. Each consecutive chunk has to list a folder with even more chunks. In total you have to do 1+2+...+100 000 = 5 000 050 000 file lookups. When you can do 1000 file lookups per millisecond (NVMe/PCIe 1 million IOPs, SATA SSD 100 000, SATA HDD <80), this is still 5000 seconds or 1.4 hours (10 000h on SATA HDD). This is just for listing the files, no transfer, checksum or anything else.

You can't ever have 100000 chunks those are never possible maximum is 10,000 in multipart. Even 100,000 lookups are minor for a filesystem. 

> @harshavardhana Please read again through the thread (especially [#7206 (comment)](https://github.com/minio/minio/issues/7206#issuecomment-505835867)). I think the agreement was reached that the coalesce at the end (tracked in #9657) is **not** the problem we are looking at here!

Coalesce is the original problem and that is the real problem, I don't know who makes that agreement. We never made any such claim.

**Comment by darkdragon-001** [comment]
> > When you have a chunk size of 1 MB, a file of 100 GB will have 100 000 chunks. The 1000th chunk (1% process) already has to list a folder with 1000 files. Each consecutive chunk has to list a folder with even more chunks. In total you have to do 1+2+...+100 000 = 5 000 050 000 file lookups. When you can do 1000 file lookups per millisecond (NVMe/PCIe 1 million IOPs, SATA SSD 100 000, SATA HDD <80), this is still 5000 seconds or 1.4 hours (10 000h on SATA HDD). This is just for listing the files, no transfer, checksum or anything else.
> 
> You can't ever have 100000 chunks those are never possible maximum is 10,000 in multipart. Even 100,000 lookups are minor for a filesystem.

The problem is not to do the lookup after the _upload_ finishes but to do it after every _chunk_! It scales with O(n²) in the current implementation.

10,000 chunks still result in 1+2+...10,000 = 50,005,000 lookups. For an SATA HDD this is still 500,050 seconds = 140 hours = 6 days. This is just for listing the files, no transfer, checksum or anything else.

**Comment by darkdragon-001** [comment]
@harshavardhana Awesome :heart_eyes: 

I just checked it and upload rate stays almost constant while it dropped by 93% before (5x speed increase)!

**Comment by acejam** [comment]
@harshavardhana I'm seeing this same problem when running with erasure coding in distributed mode.

The configuration below causes slow Multipart uploads with large files.
```[default]
region = us-east-1
s3 =
    multipart_threshold = 5MB
    multipart_chunksize = 5MB
    max_concurrent_requests = 1
    addressing_style = path
```

If I leave the above configuration in place and simply change `multipart_chunksize` to `64MB` (or anything larger than ~10-15 MB), MinIO performs significantly faster, nearly 90% in certain cases.

Is there a way to apple a similar fix here but for erasure coded setups?

**Comment by harshavardhana** [comment]
> @harshavardhana I'm seeing this same problem when running with erasure coding in distributed mode.
> 
> The configuration below causes slow Multipart uploads with large files.
> 
> ```
> region = us-east-1
> s3 =
>     multipart_threshold = 5MB
>     multipart_chunksize = 5MB
>     max_concurrent_requests = 1
>     addressing_style = path
> ```
> 
> If I leave the above configuration in place and simply change `multipart_chunksize` to `64MB` (or anything larger than ~10-15 MB), MinIO performs significantly faster, nearly 90% in certain cases.
> 
> Is there a way to apple a similar fix here but for erasure coded setups?

There is no such fix for Erasure coded setups @acejam - this issue old and let's not discuss here. Please open a new issue to discuss further.
